Put incoming vehicles in separate field. Detect incoming from other routes.

skyqrose opened the pull request:
<img width="755" alt="Screen Shot 2019-07-18 at 16 49 36" src="https://user-images.githubusercontent.com/23065557/61720350-9e76c800-ad34-11e9-9c3d-ed69ede5682f.png">


Asana Task: [Find vehicles by block that should be coming on to this route, display them in Incoming box](https://app.asana.com/0/1112935048846093/1127057236632751)

And as a part of that: [Load calendar data from GTFS](https://app.asana.com/0/397027775522669/1130527070587112), [Calculate active_blocks](https://app.asana.com/0/397027775522669/1130527070587113)

Incoming vehicles are calculated on the backend and put into a separate field in the data sent up the socket. Most of that logic has been moved into `incoming_vehicles.ex`

The active_blocks for a given time range is calculated by a linear filter over all trips. I'm not sure if that will be a problem.

Sorry for the large PR. I’ve kept meaningful commits if you’d find it easier to go commit by commit

arkadyan commented on lib/realtime/incoming_vehicles.ex:
```diff
@@ -0,0 +1,103 @@
+defmodule Realtime.IncomingVehicles do
```
I had a hard time getting a handle on the new flow of things and I'm trying to figure out how to make that concrete.

One thing that comes to mind is that a bunch of naming uses "incoming" in a way that I'm not sure clarifies as well as they could.

For one example `IncomingVehicles` really seems like it's more about taking all vehicles in and partitioning them into "on route" and "incoming" as equal partners on the way out. I'm not sure exactly what a better name here would be, `PartitionedVehicles` comes to mind but I'm not sure if that's as descriptive as it could be.

This is similarly reflected in the type `vehicles_with_incoming`. For me this doesn't really reflect the structure and intent of the data inside.

skyqrose replied:
Yeah, I agree. The way it currently is, it's like once they're partitioned, the non-incoming vehicles become the responsibility of this module, which is named as if it's just responsible for incoming vehicles, and that doesn't feel quite right.

`PartitionedVehicles` gets across the structure of the data, and feels like an appropriate name to be responsible for that structure, but doesn't really say what they're partitioned by.

I'll try to find a better name for that `%{on_route, incoming}` type.

skyqrose commented on test/gtfs/calendar_test.exs:
```diff
@@ -0,0 +1,53 @@
+      calendar_dates_txt =
+        [
+          "service_id,date,exception_type,holiday_name",
+          "from,20190102,2,",
```
I've changed `from` and `to` to `service-with-removed-date` and `service-with-added-date`.

skyqrose replied:
I renamed the services to `service-with-removed-date` and `service-with-added-date` and removed an irrelevant part of the test.

skyqrose commented on lib/gtfs/calendar.ex:
```diff
@@ -0,0 +1,131 @@
+
+  @spec from_files(binary() | nil, binary() | nil) :: t()
+  def from_files(calendar_file, calendar_dates_file) do
+    gtfs_calendar = Csv.parse(calendar_file, fn _row -> true end, &calendar_from_csv_row/1)
```
I think this case is a little different, since there are two different files, and they're isolated from the other data (meaning they don't need the list of bus route ids or anything).

I just tried moving the parsing into `data.ex`, but it meant exporting 3 functions and 2 types instead of just 1 function (`from_files` vs `gtfs_calendar_row`, `gtfs_calendar_date_row`, `calendar_from_csv_row`, `calendar_date_from_csv_row`, and `calendar_from_gtfs`).

`data.ex` is also bigger and more complicated than `calendar.ex`, and I'd rather keep logic out of it if it's reasonable.

And even with the move, `calendar.ex` still has to understand gtfs and its format, since it still needs to define the parsers.

The inconsistency is a little weird, but I think the rest of the situation makes it worth it to be inconsistent.

skyqrose commented on lib/util/time.ex:
```diff
@@ -130,6 +126,18 @@ defmodule Util.Time do
+  This is the date that roles over at midnight, not the service date.
+  """
+  @spec date_for_timestamp(timestamp()) :: Date.t()
+  def date_for_timestamp(timestamp) do
```
I expect yours to go in first. I'll rebase on top of it when that happens.

skyqrose commented on lib/gtfs/data.ex:
```diff
@@ -83,6 +83,50 @@ defmodule Gtfs.Data do
+      end_time
+      |> Util.Time.date_for_timestamp()
+
+    dates = Enum.to_list(Date.range(first_possible_service_date, last_possible_service_date))
```
I don't like using pipes when neither of the arguments feels primary. In this case, I think splitting the start and end of the range onto different lines would be worse for readability than the nested parentheses. And since the dates are the second argument of the function call on the next line, that line can't be incorporated into a longer pipeline.

arkadyan replied:
Another option is an intermediate variable—`date_range` maybe. That also helps with the "hard to read nested function calls" problem.

skyqrose commented on lib/gtfs/calendar.ex:
```diff
@@ -0,0 +1,131 @@
+
+      Enum.flat_map(dates, fn date ->
+        is_active? =
+          case Date.day_of_week(date) do
```
Good call.

skyqrose commented on lib/gtfs/data.ex:
```diff
@@ -83,6 +83,50 @@ defmodule Gtfs.Data do
+
+        Enum.filter(Map.values(trips), fn trip ->
+          trip.service_id in service_ids and
+            end_time_of_day > List.first(trip.stop_times).time and
```
Mm. I like it. I see like 7 places that do this.

skyqrose commented on test/gtfs/data_test.exs:
```diff
@@ -323,4 +323,343 @@ defmodule Gtfs.DataTest do
   end
+
+  describe "active_blocks" do
+    # @spec active_blocks(t(), Util.Time.timestamp(), Util.Time.timestamp()) :: %{Route.id() => [Block.id()]}
```
Oops. Thanks.

skyqrose commented:
@arkadyan I make a `Vehicles.for_route()` / `VehiclesForRoute` type on the backend / frontend, and pretty much everything uses that type. I like it a lot more than what was there before.

I think I've addressed everything else, too. Ready for another review.

I've rebased to take care of merge conflicts, but haven't squashed yet, so [this comparison](https://github.com/mbta/skate/compare/8b91d01...sqr-incoming) has everything past the first 3 commits that you haven't reviewed yet.

arkadyan commented on test/realtime/incoming_vehicles_test.exs:
```diff
@@ -0,0 +1,114 @@
+defmodule Realtime.VehiclesTest do
```
Looks like the filename needs to be updated to match the new module name.

arkadyan commented on test/realtime/incoming_vehicles_test.exs:
```diff
@@ -0,0 +1,114 @@
+             }
+    end
+
+    test "includes vehicles incoming onto a new route in their new route" do
```
It looks like it will also include vehicles that are coming from no route just fine, but could you please add a test demonstrating that case? I know it's a little didactic, but I could imagine it being a distinction worth highlighting.

skyqrose replied:
I didn't even realize route_id could be assigned nil. Definitely worth a test.

arkadyan commented on lib/gtfs/block.ex:
```diff
@@ -32,9 +31,7 @@ defmodule Gtfs.Block do
-    |> Map.get(:stop_times)
-    |> List.first()
-    |> Map.get(:time)
+    |> Trip.start_time()
```
👍

arkadyan commented on lib/realtime/server.ex:
```diff
@@ -32,57 +29,63 @@ defmodule Realtime.Server do
   end
 
+  @spec initial_state() :: state()
+  defp initial_state(), do: %{}
```
What's the benefit of breaking this out into a function?

skyqrose replied:
Oops, this was left over from a previous change.

arkadyan commented on lib/realtime/server.ex:
```diff
@@ -9,13 +9,10 @@ defmodule Realtime.Server do
-  @type state :: %{
-          vehicles: vehicles_by_route_id()
-        }
+  @typep state :: Route.by_id(Vehicles.for_route())
```
Do the server tests need to be updated to reflect all these changes?

skyqrose replied:
Ooh, thanks, yeah. I don't know how they were even passing.

arkadyan commented on lib/gtfs/data.ex:
```diff
@@ -68,6 +82,54 @@ defmodule Gtfs.Data do
+    active_services = Map.take(calendar, dates)
+
+    active_trips =
+      Enum.flat_map(active_services, fn {date, service_ids} ->
```
It's not necessary, but I could see this getting refactored out as a function on `Trip`.

skyqrose replied:
Hmm.. I just tried but the service / time range comparisons are kind of specific to what's going on in this file so it didn't really feel like an improvement.

skyqrose commented:
Ready for final review.
